File: DockerSubmission/submission1-Python/driver.py
#!/usr/bin/env python3

import numpy as np, os, sys
from get_sepsis_score import get_sepsis_score

def load_challenge_data(file):
    with open(file, 'r') as f:
        header = f.readline().strip()
        column_names = header.split('|')
        data = np.loadtxt(f, delimiter='|')

    # ignore SepsisLabel column if present
    if column_names[-1] == 'SepsisLabel':
        column_names = column_names[:-1]
        data = data[:, :-1]

    return data

def save_challenge_predictions(file, scores, labels):
    with open(file, 'w') as f:
        f.write('PredictedProbability|PredictedLabel\n')
        for (s, l) in zip(scores, labels):
            f.write('%g|%d\n' % (s, l))

if __name__ == '__main__':
    # Parse arguments.
    if len(sys.argv) != 3 and len(sys.argv) != 4:
        raise Exception('Include the input and output directories as arguments, e.g., python driver.py input output.')

    input_directory = sys.argv[1]
    output_directory = sys.argv[2]

    if len(sys.argv) == 3 or sys.argv[0] == '0' or sys.argv[0] == 'False':
        enforce_causality = False
    else:
        enforce_causality = True

    # Find files.
    files = []
    for f in os.listdir(input_directory):
        if os.path.isfile(os.path.join(input_directory, f)) and not f.lower().startswith('.') and f.lower().endswith('psv'):
            files.append(f)

    # Iterate over files.
    for file in files:
        # Load data.
        input_file = os.path.join(input_directory, file)
        data = load_challenge_data(input_file)

        # Make predictions.
        if not enforce_causality:
            scores, labels = get_sepsis_score(data)
        else:
            num_records = len(data)
            scores = np.zeros(num_records)
            labels = np.zeros(num_records)
            for t in range(num_records):
                current_data = data[:t+1]
                current_scores, current_labels = get_sepsis_score(current_data)
                scores[t] = current_scores[t]
                labels[t] = current_labels[t]

        # Save results.
        output_file = os.path.join(output_directory, file)
        save_challenge_predictions(output_file, scores, labels)

File: DockerSubmission/submission1-R/driver.R
#!/usr/bin/Rscript

source("get_sepsis_score.R")

load_challenge_data = function(file){
    data = data.matrix(read.csv(file, sep='|'))
    column_names = colnames(data)

    # ignore SepsisLabel column if present
    if (column_names[ncol(data)] == 'SepsisLabel'){
        data = data[, 1:ncol(data)-1]
    }

    return(data)
}

save_challenge_predictions = function(file, predictions){
    colnames(predictions) = c('PredictedProbability', 'PredictedLabel')
    write.table(predictions, file = file, sep = '|', quote = FALSE, row.names = FALSE)
}

# Parse arguments.
args = commandArgs(trailingOnly=TRUE)
if (length(args) != 2 && length(args) != 3){
    stop('Include the input and output directories as arguments, e.g., Rscript driver.r input output.')
}

input_directory = args[1]
output_directory = args[2]

if (length(args) == 2 || args[3] == '0' || args[3] == 'F' || args[3] == 'FALSE'){
    enforce_causality = FALSE
} else{
    enforce_causality = TRUE
}

# Find files.
files = c()
for (f in list.files(input_directory)){
    if (file.exists(file.path(input_directory, f)) && nchar(f) >= 3 && substr(f, 1, 1) != '.' && substr(f, nchar(f)-2, nchar(f)) == 'psv'){
        files = c(files, f)
    }
}

if (!dir.exists(output_directory)){
    dir.create(output_directory)
}

# Iterate over files.
for (file in files){
    # Load data.
    input_file = file.path(input_directory, file)
    data = load_challenge_data(input_file)

    # Make predictions.
    if (!enforce_causality){
        predictions = get_sepsis_score(data)
    } else{
        num_rows = nrow(data)
        predictions = matrix(, num_rows, 2)
        for (t in 1 : num_rows){
            current_data = matrix(data[1 : t, ], t, ncol(data))
            current_predictions = get_sepsis_score(current_data)
            predictions[t, ] = current_predictions[t, ]
        }
    }

    # Save results.
    output_file = file.path(output_directory, file)
    save_challenge_predictions(output_file, predictions)
}
